app/src/test/java/de/westnordost/streetcomplete/data/user/statistics/StatisticsParserTest.kt

```diff
@@ -1,80 +0,0 @@
-package de.westnordost.streetcomplete.data.user.statistics
-
-import kotlinx.datetime.Instant
-import kotlinx.datetime.LocalDate
-import kotlin.test.Test
-import kotlin.test.assertEquals
-
-class StatisticsParserTest {
```

Comment:
You removed this test. I think it is better if the `StatisticsParserTest` remained and tested the statistics parser, while the `StatisticsDownloaderTest` *only* tests the downloading. I.e. mocks an own `StatisticsParser` that just always returns a valid `Statistics` or whatever is required by the specific test.

Reply:
Yeah, I think this is my opinions on software cohesion leaking out. 😓 In my head the code to communicate with an API, and the code to convert from the APIs representation to application domain representation make a lot of sense to stay together. I can't imagine the parser being used in any context other than with that API. I also suspected that the only reason for separate `StatisticsParser` and `StatisticsDownloader` classes was because the "parsing" behaviour was hard to test previously, and I'd actually push for combining the two classes 🤣 .

I'm happy to put my thoughts on software design in a box and split out the tests later this week.

Reply:
But wouldn't you agree that as long as the classes are separate, the tests should also be separate? Sure, they are closely related, that's why they are in the same package. There are several levels of how functionality that is closely related/interdependent can be separated - in the same package, in the same file, in the same class, in the same function, ...and, well, I guess different philosophies do exist. Separating parser, (generator,) and IO is however done quite consistently throughout the project IIRC and I like the code base to stay as consistent as possible.

Nevertheless, I would be interested to know where you would draw the line, when to separate and when not to separate, and how to separate, if you like to share that with me.

Reply:
> But wouldn't you agree that as long as the classes are separate, the tests should also be separate?

I'm more of the opinion that test design and code design should be free to diverge independently. Tests exist to test behaviours of the system, not classes. In some cases, I may want to write several classes worth of tests, for a single production class. In some cases I may be able to define the behaviour of the software within one test class, but want to use several different production classes to implement said behaviour. See also [this video](https://youtu.be/EZ05e7EMOLM?feature=shared&t=1350).

> Separating parser, (generator,) and IO is however done quite consistently throughout the project IIRC and I like the code base to stay as consistent as possible.

Agree with consistency being important 👍🏻

> Nevertheless, I would be interested to know where you would draw the line, when to separate and when not to separate, and how to separate, if you like to share that with me.

When I'm thinking about testing, I'm usually thinking first about what boundaries do I have around my softwares components or want to have around my components. In this case, I see:
1. An API which provides a very hard boundary between once software component, and the mobile app. This being a boundary or not is outside of our control.
2. We also have this `data class Statistics` which somebody has sensibly designed, in a way where it is used across multiple different components of the app, and generated from one place. Allows statistics to be cached locally, or even calculated locally in some future without needing to change the design of other components. A useful boundary.

To me, with this mindset of software boundaries I'd then be wanting good confidence that a particular end to end boundary works, and being lazy I'd default to testing it all together as one. See also [Layered Architecture](https://ddd-practitioners.com/home/glossary/layered-architecture/).

I don't see the StatisticsParser as a valuable boundary. `StatisticsDownloader` has an explicit dependency on `StatisticsParser` via the constructor parameter. `StatisticsParser` (as far as I can imagine) would only ever be used by a `StatisticsDownloader` instance. These classes will likely always appear as a pair. Friends forever 😄.

Reply:
If tests are not coupled to classes, it becomes harder to ascertain the test coverage on a glance, though. But yeah, I recognize that this is a very imprecise measure anyway and there are better tools for that.

I've watched the talk (till min 40). Interesting. Though in general he starts from a pretty dogmatic tests-first / "100% test coverage" uhm... hellhole and I never bought into TDD that much that I could relate or experienced any of the issues he described with TDD - TDD has been the exception rather than the rule for me.

Anyway, what you write makes sense. In a nutshell, if you find it cleaner to merge them into one class, I'd be fine with that, but IMO if you do, it should happen both for the tests and the production code, i.e. the parser should not be injectable in the `StatisticsDownloader` in that case but be part of that class.

We say we make dependencies injectable for easier testing. But when the tests should ideally not concern themselves with interactions within a "module", i.e. interactions that are implementation details from a more ehm... birds eye view the tests should ideally be written from, then it also makes no sense to have these "dependencies" injectable.

I fear, if one follows that paradigma to the extreme again (like the speaker with TDD apparently at the beginning), one could end up with quite monolithic class giants that actually become more complex to test because they do many things at the same time.

I think, the beauty of tearing apart into separate classes functionalities even within one use case / module / ... is to make sure that ideally one class / file only does one thing, and that one thing can easily be tested. Whether or not that thing is only used at one place within the same module seems not to be too relevant to the test itself.

Anyway, I don't have such a strong opinion about it and I feel we are not actually that far away from each other opinion-wise. The `Statistics*` is a small thing and whether it is done one way or another ultimately doesn't really change anything. Where exactly the line is drawn (=when something is an own class, an own function, ...) is always subject of consideration